### PR TITLE
[Refactor] Refactor unit test structure

### PR DESCRIPTION
**Pull request by congee524**
Refactor unittest structure according to design document

### CODE REVIEW

**Comment by innerlee** [comment]
What's the status now, is it ready for review?

**Comment by congee524** [comment]
yes. And we will then raise another pull request to extract the common function in mmcv

**Comment by dreamerlin** [comment]
Here are some suggestions to discuss:
1. Data directory in `tests/data` should remove prefix `test` in the name of theirs, and we can create a `tests/data/annotations` directory to hold the annotation files.
2. How about creating a `base.py` under `tests/test_data/test_datasets`, where we can define a `class BaseTestDataset(cls)` with `def setup_class(cls)` holding some basic info like `cls.test_data_directory` and `cls.test_anno_directory` to avoid dupilcated `../data` for certain dataset by inheriting it. And this thought can be applied to other parts in the unittest.
3. Same as 2. some frequently used function like `generate_demo_inputs`, `_demo_inputs` , etc. can be integrated and put in a `base.py` which can be inherit by other modules. (Note, these functions may be not able to put in MMCV due to the difference of the data among different repos)
4. Some filenames should be plural, like `test_common_module -> test_common_modules`.
5. Some file position is not correct. `test_inference.py` should be in `test_runtime` since it is more about the pipeline of the repo rather than a certain model. `test_gradcam.py` should be put in `test_models` since it is more about the model not the utils. `test_onnx.py` should be put in `test_utils.py` since it is more about a onnx conversion tools. etc.
6. To reduce some duplicated code, you can refer to [this](https://github.com/open-mmlab/mmcv/blob/master/tests/test_cnn/test_wrappers.py), and try to use `pytest.mark.parametrize` for some codes.

What do you think ? @congee524 @innerlee

**Comment by congee524** [comment]
> Here are some suggestions to discuss:
> 
> 1. Data directory in `tests/data` should remove prefix `test` in the name of theirs, and we can create a `tests/data/annotations` directory to hold the annotation files.
> 2. How about creating a `base.py` under `tests/test_data/test_datasets`, where we can define a `class BaseTestDataset(cls)` with `def setup_class(cls)` holding some basic info like `cls.test_data_directory` and `cls.test_anno_directory` to avoid dupilcated `../data` for certain dataset by inheriting it. And this thought can be applied to other parts in the unittest.
> 3. Same as 2. some frequently used function like `generate_demo_inputs`, `_demo_inputs` , etc. can be integrated and put in a `base.py` which can be inherit by other modules. (Note, these functions may be not able to put in MMCV due to the difference of the data among different repos)
> 4. Some filenames should be plural, like `test_common_module -> test_common_modules`.
> 5. Some file position is not correct. `test_inference.py` should be in `test_runtime` since it is more about the pipeline of the repo rather than a certain model. `test_gradcam.py` should be put in `test_models` since it is more about the model not the utils. `test_onnx.py` should be put in `test_utils.py` since it is more about a onnx conversion tools. etc.
> 6. To reduce some duplicated code, you can refer to [this](https://github.com/open-mmlab/mmcv/blob/master/tests/test_cnn/test_wrappers.py), and try to use `pytest.mark.parametrize` for some codes.
> 
> What do you think ? @congee524 @innerlee

about the first point, does it means `test_activity_features -> activity_features` ?
Does the anotation files means hte anotation files of training, validation and testing?

**Comment by dreamerlin** [comment]
> test_activity_features -> activity_features

like `test_activity_features -> activity_features`

**Comment by innerlee** [comment]
Is this ready for review & merge?

**Comment by congee524** [comment]
> Is this ready for review & merge?

I'm waiting for releasing of mmcv to apply its latest common testing function

File: tests/test_data/test_datasets/test_audio_dataset.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+        with pytest.raises(KeyError):
+            # unsupported metric
+            audio_dataset.evaluate([0] * len(audio_dataset), metrics='iou')
+
+        # evaluate top_k_accuracy and mean_class_accuracy metric
+        results = [np.array([0.1, 0.5, 0.4])] * 2
+        eval_result = audio_dataset.evaluate(
+            results, metrics=['top_k_accuracy', 'mean_class_accuracy'])
+        assert set(eval_result.keys()) == set(
```
Comment (dreamerlin) [inline tests/test_data/test_datasets/test_audio_dataset.py:76]: `set(eval_result)` is ok

File: tests/test_data/test_datasets/test_rawframe_dataset.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        with pytest.raises(KeyError):
+            # unsupported metric
+            rawframe_dataset.evaluate(
+                [0] * len(rawframe_dataset), metrics='iou')
+
+        # evaluate top_k_accuracy and mean_class_accuracy metric
+        results = [np.array([0.1, 0.5, 0.4])] * 2
+        eval_result = rawframe_dataset.evaluate(
+            results, metrics=['top_k_accuracy', 'mean_class_accuracy'])
+        assert set(eval_result.keys()) == set(
```
Comment (dreamerlin) [inline tests/test_data/test_datasets/test_rawframe_dataset.py:163]: same

File: tests/test_data/test_datasets/test_repeat_dataset.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    def test_repeat_dataset(self):
+        rawframe_dataset = RawframeDataset(self.frame_ann_file,
+                                           self.frame_pipeline,
+                                           self.data_prefix)
+        repeat_dataset = RepeatDataset(rawframe_dataset, 5)
+        assert len(repeat_dataset) == 10
+        result_a = repeat_dataset[0]
+        result_b = repeat_dataset[2]
+        assert set(result_a.keys()) == set(result_b.keys())
```
Comment (dreamerlin) [inline tests/test_data/test_datasets/test_repeat_dataset.py:17]: same

File: tests/test_data/test_datasets/test_ssn_dataset.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+        # SSN Dataset not in test mode
+        proposal_dataset = SSNDataset(
+            self.proposal_ann_file,
+            self.proposal_pipeline,
+            self.proposal_train_cfg,
+            self.proposal_test_cfg,
+            data_prefix=self.data_prefix)
+        result = proposal_dataset[0]
+        assert self.check_keys_contain(result.keys(), target_keys)
```
Comment (dreamerlin) [inline tests/test_data/test_datasets/test_ssn_dataset.py:26]: can this be replaced with MMCV function ?

File: tests/test_data/test_datasets/test_video_dataset.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+        with pytest.raises(KeyError):
+            # unsupported metric
+            video_dataset.evaluate([0] * len(video_dataset), metrics='iou')
+
+        # evaluate top_k_accuracy and mean_class_accuracy metric
+        results = [np.array([0.1, 0.5, 0.4])] * 2
+        eval_result = video_dataset.evaluate(
+            results, metrics=['top_k_accuracy', 'mean_class_accuracy'])
+        assert set(eval_result.keys()) == set(
```
Comment (dreamerlin) [inline tests/test_data/test_datasets/test_video_dataset.py:79]: same

**Comment by dreamerlin** [comment]
global search for some common comments

**Comment by dreamerlin** [comment]
It seems the coverage drops ?

**Comment by dreamerlin** [comment]
> It seems the coverage drops ?

`test_tem.py` is empty. May need to check if all the unit tests are moved correctly

**Review by dreamerlin — APPROVED** [review]

**Comment by dreamerlin** [comment]
changelog is needed @congee524

**Comment by kennymckormick** [comment]
I think `est_tem_results` should be removed, is it a typo?

**Comment by congee524** [comment]
> I think `est_tem_results` should be removed, is it a typo?

yeah, and I have removed it

File: tests/test_data/test_pipelines/test_augmentation/base.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        return np.array_equal(
+            np.array(origin_imgs)[:, top:bottom, left:right, :],
+            np.array(result_imgs))
+
+    if result_bbox.ndim == 1:
+        return check_single_crop(origin_imgs, result_imgs, result_bbox)
+    elif result_bbox.ndim == 2:
+        num_batch = len(origin_imgs)
+        for i, bbox in enumerate(result_bbox):
+            if num_crops == 10:
```
[the remark this answers is not in the document]
Comment (congee524) [reply tests/test_data/test_pipelines/test_augmentation/base.py:25]: We don't need to deal with the case. Flip_crop only appears in TenCrop

**Comment by congee524** [comment]
kindly ping :p @innerlee @SuX97

File: tests/data/annotations/frame_test_list.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+imgs 5 127
```
Comment (SuX97) [inline tests/data/annotations/frame_test_list.txt:1]: `frame` -> `rawframe`

File: tests/data/annotations/rawvideo_test_anno.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+[
+  {
+    "video_dir": "rawvideo_dataset",
```
Comment (SuX97) [inline tests/data/annotations/rawvideo_test_anno.json:3]: What's the difference between rawvideo and video? I think it should be consistent of dataset types.

File: tests/test_data/test_pipelines/test_augmentation/base.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        for i in range(n):
+            for j in range(h):
+                for k in range(w):
+                    for channel in range(c):
+                        if result_imgs[i][j, k, channel] != origin_imgs[i][j, w - 1 - k, channel]:  # noqa:E501
+                            return False
+        # yapf: enable
+    else:
+        # yapf: disable
+        for i in range(n):
```
Comment (SuX97) [inline tests/test_data/test_pipelines/test_augmentation/base.py:59]: Can we just use `numpy.fliplr`? Maybe faster.

File: tests/test_data/test_pipelines/test_augmentation/test_color.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+from numpy.testing import assert_array_equal
+
+from mmaction.datasets.pipelines import ColorJitter
+
+
+class TestColor:
+
+    def test_color_jitter(self):
+        imgs = list(
+            np.random.randint(0, 255, size=(3, 240, 320, 3), dtype=np.uint8))
```
Comment (SuX97) [inline tests/test_data/test_pipelines/test_augmentation/test_color.py:12]: Do we really need such big img?

File: tests/test_data/test_pipelines/test_loading/base.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
+import os.path as osp
+
+import mmcv
+import numpy as np
+
+
+class ExampleSSNInstance:
```
Comment (SuX97) [inline tests/test_data/test_pipelines/test_loading/base.py:7]: This class is model specified, should not in base. Is it possible to directly import from `ssn_dataset.py`?

File: tests/test_models/test_common_modules/test_base_head.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    assert 'loss_cls' in losses.keys()
+    assert losses.get('loss_cls') > 0, 'cls loss should be non-zero'
+
+    head = ExampleHead(3, 400, dict(type='CrossEntropyLoss', loss_weight=2.0))
+
+    cls_scores = torch.rand((3, 4))
+    # When truth is non-empty then cls loss should be nonzero for random inputs
+    gt_labels = torch.LongTensor([2] * 3).squeeze()
+    losses = head.loss(cls_scores, gt_labels)
+    assert 'loss_cls' in losses.keys()
```
Comment (SuX97) [inline tests/test_models/test_common_modules/test_base_head.py:31]: Maybe not use the function from `mmcv.utils` like `check_contain_keys`

File: tests/test_models/test_common_modules/test_base_recognizers.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        wrong_test_cfg = dict(average_clips='softmax')
+        recognizer = ExampleRecognizer(None, wrong_test_cfg)
+        recognizer.average_clip(cls_score)
+
+    with pytest.raises(ValueError):
+        # Label should not be None
+        recognizer = ExampleRecognizer(None, None)
+        recognizer(torch.tensor(0))
+
+    # average_clips='score'
```
Comment (SuX97) [inline tests/test_models/test_common_modules/test_base_recognizers.py:48]: Just a reminder: `average_clips=None` is not tested, this usage is rather important. Maybe fix it in other PR when time permits.

**Review by SuX97 — COMMENTED** [review]
1. May need to check if those common check functions moved to `mmcv.utils` are actually being used.
2. How to set a threshold to split the unittest?

File: tests/data/annotations/rawvideo_test_anno.json (hunk 8 above)
Comment (congee524) [reply tests/data/annotations/rawvideo_test_anno.json:3]: RawVideo dataset is used in the Project OmniSource. They have some differences.

File: .gitignore [hunk 14]
```diff
@@ -119,3 +119,10 @@ work_dirs/
 
 # Profile
 *.prof
+
+# unignore some data file in tests/data
+!tests/data/**/*.pkl
+!tests/data/**/*.pkl.json
+!tests/data/**/*.log.json
+!tests/data/**/*.pth
+!tests/data/**/*.prof
```
Comment (innerlee) [inline .gitignore:128]: is profile file needed anywhere?

File: docs/changelog.md [hunk 15]
```diff
@@ -11,6 +11,7 @@
 **Improvements**
 
 - Support setting `max_testing_views` for extremely large models to save GPU memory used ([#511](https://github.com/open-mmlab/mmaction2/pull/511))
+- Refactor unit test structure
```
Comment (innerlee) [inline docs/changelog.md:14]: write pr link

File: tests/data/annotations/hvu_frame_test_anno.json [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+      ],
+      "action": [
+        180
+      ],
+      "scene": [
+        206
+      ]
+    }
+  }
+]
```
Comment (innerlee) [inline tests/data/annotations/hvu_frame_test_anno.json:58]: end the file with \newline

**Comment by congee524** [comment]
> 1. May need to check if those common check functions moved to `mmcv.utils` are actually being used.
> 2. How to set a threshold to split the unittest?

I have checked it all before, but there may be some omissions. We may fix them in future PR If there exist some omissions.

There is no threshold for the time being. In fact, I have some confusion about `test_localizers` and `test_backbones`

**Review by innerlee — APPROVED** [review]

**Pull request merged**